snmp/datadog_checks/snmp/snmp.py

```diff
@@ -280,14 +283,14 @@ def _profile_for_sysobject_oid(self, sys_object_oid):
         """
         Return the most specific profile that matches the given sysObjectID.
         """
-        profiles = [profile for oid, profile in self.profiles_by_oid.items() if fnmatch.fnmatch(sys_object_oid, oid)]
+        matched_oids = [
+            (oid, profile) for oid, profile in self.profiles_by_oid.items() if fnmatch.fnmatch(sys_object_oid, oid)
+        ]
 
-        if not profiles:
+        if not matched_oids:
             raise ConfigurationError('No profile matching sysObjectID {}'.format(sys_object_oid))
 
-        return max(
-            profiles, key=lambda profile: oid_pattern_specificity(self.profiles[profile]['definition']['sysobjectid'])
-        )
+        return max(matched_oids, key=lambda oid: oid_pattern_specificity(oid[0]))[1]
```

Comment:
I think if possible, we should avoid using indexes like here `oid[0]` and `oid[1]`, it can be hard to read.

An alternative could be:

```python
    def _profile_for_sysobject_oid(self, sys_object_oid):
        # just filter and return same kind of dict
        matched_profiles_by_oid = {
            self.profiles_by_oid[oid] for oid in self.profiles_by_oid if fnmatch.fnmatch(sys_object_oid, oid)
        }

        if not matched_profiles_by_oid:
            raise ConfigurationError('No profile matching sysObjectID {}'.format(sys_object_oid))

        max_oid = max(matched_profiles_by_oid.keys(), key=lambda oid: oid_pattern_specificity(oid))
        return matched_profiles_by_oid[max_oid]
```

Reply:
Agreed with Alex^, but I think he meant

```python
matched_profiles_by_oid = {
    oid: self.profiles_by_oid[oid] for oid in self.profiles_by_oid if fnmatch.fnmatch(sys_object_oid, oid)
}